File: src/credit_audit/interventions/monotone.py
"""Absolute, boundary-targeted monotonicity plans for the synthetic policy."""

from __future__ import annotations

from decimal import ROUND_CEILING, ROUND_FLOOR, Decimal

from credit_audit.ids import applicant_content_id, cluster_id_for, content_id
from credit_audit.interventions.apply import PairPlan, make_pair_plan
from credit_audit.policy.loader import Policy
from credit_audit.policy.oracle import evaluate
from credit_audit.types import (
    Applicant,
    DecisionOutcome,
    Family,
    Frozen,
    FrozenDict,
    InterventionSpec,
    Layer,
    Relation,
)

CHECK_INCOME = "monotonicity.income_increase"
CHECK_CREDIT_SCORE = "monotonicity.credit_score_increase"
CHECK_DTI = "monotonicity.dti_increase"
CHECK_MINOR_DELINQUENCY = "monotonicity.minor_delinquency_increase"
CHECK_MAJOR_DELINQUENCY = "monotonicity.major_delinquency_increase"
CHECK_INCOME_ANALYTIC = "monotonicity.income_30k_to_40k"


class MonotonicityCase(Frozen):
    applicant_id: str
    check: str
    target_rule_id: str
    expected_relation: Relation
    pair_id: str
    cluster_id: str
    plan: PairPlan | None = None
    inapplicable_reason: str | None = None


def _intervention(
    *,
    applicant: Applicant,
    check: str,
    arm: str,
    targets: dict[str, object],
    relation: Relation,
) -> InterventionSpec:
    payload = {
        "applicant_id": applicant.applicant_id,
        "check": check,
        "arm": arm,
        "targets": targets,
    }
    return InterventionSpec(
        intervention_id=content_id(payload),
        family=Family.MONOTONE,
        name=f"{check}:{arm}",
        layer=Layer.FACTS,
        target_field=next(iter(targets)),
        direction="set",
        expected_relation=relation,
        params=FrozenDict({"targets": targets}),
    )


def _inapplicable(
    applicant: Applicant,
    check: str,
    rule_id: str,
    relation: Relation,
    reason: str,
) -> MonotonicityCase:
    pair_id = content_id(
        {
            "applicant_id": applicant.applicant_id,
            "applicant_content_id": applicant_content_id(applicant),
            "check": check,
            "inapplicable": reason,
        }
    )
    return MonotonicityCase(
        applicant_id=applicant.applicant_id,
        check=check,
        target_rule_id=rule_id,
        expected_relation=relation,
        pair_id=pair_id,
        cluster_id=cluster_id_for(applicant),
        inapplicable_reason=reason,
    )


def _build_case(
    *,
    applicant: Applicant,
    policy: Policy,
    check: str,
    rule_id: str,
    relation: Relation,
    base_targets: dict[str, object],
    cf_targets: dict[str, object],
    expected_base_breaches: tuple[str, ...],
    expected_cf_breaches: tuple[str, ...],
) -> MonotonicityCase:
    base_spec = _intervention(
        applicant=applicant,
        check=check,
        arm="base",
        targets=base_targets,
        relation=relation,
    )
    cf_spec = _intervention(
        applicant=applicant,
        check=check,
        arm="counterfactual",
        targets=cf_targets,
        relation=relation,
    )
    plan = make_pair_plan(
        applicant=applicant,
        check=check,
        family=Family.MONOTONE,
        relation=relation,
        base_arm_id=f"{check}:base",
        cf_arm_id=f"{check}:counterfactual",
        base_interventions=(base_spec,),
        cf_interventions=(cf_spec,),
    )
    base = plan.base.materialize().applicant
    cf = plan.cf.materialize().applicant
    base_eval = evaluate(base.facts, policy)
    cf_eval = evaluate(cf.facts, policy)
    if base_eval.breached_rule_ids != expected_base_breaches:
        return _inapplicable(
            applicant,
            check,
            rule_id,
            relation,
            "base arm is confounded: expected breached rules "
            f"{expected_base_breaches!r}, observed {base_eval.breached_rule_ids!r}",
        )
    if cf_eval.breached_rule_ids != expected_cf_breaches:
        return _inapplicable(
            applicant,
            check,
            rule_id,
            relation,
            "counterfactual arm is confounded: expected breached rules "
            f"{expected_cf_breaches!r}, observed {cf_eval.breached_rule_ids!r}",
        )
    return MonotonicityCase(
        applicant_id=applicant.applicant_id,
        check=check,
        target_rule_id=rule_id,
        expected_relation=relation,
        pair_id=plan.pair_id,
        cluster_id=plan.cluster_id,
        plan=plan,
    )


def _dti_debt_target(monthly_income_cents: int, ratio: Decimal, *, above: bool) -> int:
    if monthly_income_cents <= 0:
        raise ValueError("a positive monthly income is required to construct a DTI boundary")
    rounding = ROUND_CEILING if above else ROUND_FLOOR
    return int((ratio * Decimal(monthly_income_cents)).to_integral_value(rounding=rounding))


def build_monotonicity_cases(
    applicant: Applicant,
    policy: Policy,
) -> tuple[MonotonicityCase, ...]:
    """Build the five predeclared, policy-boundary monotonicity contrasts.

    The selected source profile must be oracle-clean.  Each arm is then checked again
    against the oracle, and any cross-rule side effect is represented as
    ``INAPPLICABLE`` by the scorer rather than silently entering a causal denominator.
    """

    definitions = (
        (CHECK_INCOME, "min_annual_income", Relation.NONDECREASING),
        (CHECK_CREDIT_SCORE, "min_credit_score", Relation.NONDECREASING),
        (CHECK_DTI, "max_dti", Relation.NONINCREASING),
        (CHECK_MINOR_DELINQUENCY, "max_minor_delinquencies", Relation.NONINCREASING),
        (CHECK_MAJOR_DELINQUENCY, "max_major_delinquencies", Relation.NONINCREASING),
    )
    source_eval = evaluate(applicant.facts, policy)
    if source_eval.outcome is not DecisionOutcome.APPROVE:
        return tuple(
            _inapplicable(
                applicant,
                check,
                rule_id,
                relation,
                "source profile is not oracle-approved",
            )
            for check, rule_id, relation in definitions
        )

    income_rule = policy.rule("min_annual_income")
    score_rule = policy.rule("min_credit_score")
    dti_rule = policy.rule("max_dti")
    minor_rule = policy.rule("max_minor_delinquencies")
    major_rule = policy.rule("max_major_delinquencies")

    income_low = int(income_rule.threshold - income_rule.margin_unit)
    income_high = int(income_rule.threshold + income_rule.margin_unit)
    score_low = int(score_rule.threshold - score_rule.margin_unit)
    score_high = int(score_rule.threshold + score_rule.margin_unit)
    dti_low = dti_rule.threshold - dti_rule.margin_unit
    dti_high = dti_rule.threshold + dti_rule.margin_unit
    debt_low = _dti_debt_target(applicant.facts.monthly_income_cents, dti_low, above=False)
    debt_high = _dti_debt_target(applicant.facts.monthly_income_cents, dti_high, above=True)
    minor_low = int(minor_rule.threshold - minor_rule.margin_unit)
    minor_high = int(minor_rule.threshold + minor_rule.margin_unit)
    major_low = max(0, int(major_rule.threshold))
    major_high = int(major_rule.threshold + major_rule.margin_unit)

    return (
        _build_case(
            applicant=applicant,
            policy=policy,
            check=CHECK_INCOME,
            rule_id="min_annual_income",
            relation=Relation.NONDECREASING,
            base_targets={"annual_income_cents": income_low},
            cf_targets={"annual_income_cents": income_high},
            expected_base_breaches=("min_annual_income",),
            expected_cf_breaches=(),
        ),
        _build_case(
            applicant=applicant,
            policy=policy,
            check=CHECK_CREDIT_SCORE,
            rule_id="min_credit_score",
            relation=Relation.NONDECREASING,
            base_targets={"credit_score": score_low},
            cf_targets={"credit_score": score_high},
            expected_base_breaches=("min_credit_score",),
            expected_cf_breaches=(),
        ),
        _build_case(
            applicant=applicant,
            policy=policy,
            check=CHECK_DTI,
            rule_id="max_dti",
            relation=Relation.NONINCREASING,
            base_targets={"monthly_debt_cents": debt_low},
            cf_targets={"monthly_debt_cents": debt_high},
            expected_base_breaches=(),
            expected_cf_breaches=("max_dti",),
        ),
        _build_case(
            applicant=applicant,
            policy=policy,
            check=CHECK_MINOR_DELINQUENCY,
            rule_id="max_minor_delinquencies",
            relation=Relation.NONINCREASING,
            base_targets={"delinq_30d_24m": minor_low, "delinq_60d_24m": 0},
            cf_targets={"delinq_30d_24m": minor_high, "delinq_60d_24m": 0},
            expected_base_breaches=(),
            expected_cf_breaches=("max_minor_delinquencies",),
        ),
        _build_case(
            applicant=applicant,
            policy=policy,
            check=CHECK_MAJOR_DELINQUENCY,
            rule_id="max_major_delinquencies",
            relation=Relation.NONINCREASING,
            base_targets={"delinq_90p_24m": major_low},
            cf_targets={"delinq_90p_24m": major_high},
            expected_base_breaches=(),
            expected_cf_breaches=("max_major_delinquencies",),
        ),
    )


def build_analytic_income_case(
    applicant: Applicant,
    policy: Policy,
) -> MonotonicityCase:
    """The planted-control fixture: exactly $30,000 -> $40,000 annual income."""

    if evaluate(applicant.facts, policy).outcome is not DecisionOutcome.APPROVE:
        return _inapplicable(
            applicant,
            CHECK_INCOME_ANALYTIC,
            "min_annual_income",
            Relation.NONDECREASING,
            "source profile is not oracle-approved",
        )
    return _build_case(
        applicant=applicant,
        policy=policy,
        check=CHECK_INCOME_ANALYTIC,
        rule_id="min_annual_income",
        relation=Relation.NONDECREASING,
        base_targets={"annual_income_cents": 3_000_000},
        cf_targets={"annual_income_cents": 4_000_000},
        expected_base_breaches=(),
        expected_cf_breaches=(),
    )


__all__ = [
    "CHECK_CREDIT_SCORE",
    "CHECK_DTI",
    "CHECK_INCOME",
    "CHECK_INCOME_ANALYTIC",
    "CHECK_MAJOR_DELINQUENCY",
    "CHECK_MINOR_DELINQUENCY",
    "MonotonicityCase",
    "build_analytic_income_case",
    "build_monotonicity_cases",
]

File: web/lib/audit/interventions/monotone.ts
import { Applicant, Relation, Family, InterventionSpec } from "../records";
import { policy } from "../policy";
import { evaluate } from "../oracle";
import { PairPlan, makePairPlan, clusterIdFor, applyInterventions } from "./apply";
import { contentId } from "../ids";

export const CHECK_INCOME = "monotonicity.income_increase";
export const CHECK_CREDIT_SCORE = "monotonicity.credit_score_increase";
export const CHECK_DTI = "monotonicity.dti_increase";
export const CHECK_MINOR_DELINQUENCY = "monotonicity.minor_delinquency_increase";
export const CHECK_MAJOR_DELINQUENCY = "monotonicity.major_delinquency_increase";
export const CHECK_INCOME_ANALYTIC = "monotonicity.income_30k_to_40k";

export type MonotonicityCase = Readonly<{
  applicant_id: string;
  check: string;
  target_rule_id: string;
  expected_relation: Relation;
  pair_id: string;
  cluster_id: string;
  plan: PairPlan | null;
  inapplicable_reason: string | null;
}>;

function inapplicable(
  applicant: Applicant,
  check: string,
  ruleId: string,
  relation: Relation,
  reason: string
): MonotonicityCase {
  return Object.freeze({
    applicant_id: applicant.applicant_id,
    check,
    target_rule_id: ruleId,
    expected_relation: relation,
    pair_id: contentId({ check, applicant_id: applicant.applicant_id, inapplicable: true }),
    cluster_id: clusterIdFor(applicant),
    plan: null,
    inapplicable_reason: reason,
  });
}

function interventionSpec(
  applicant: Applicant,
  check: string,
  arm: string,
  targets: Record<string, any>,
  relation: Relation
): InterventionSpec {
  const targetField = Object.keys(targets)[0];
  return Object.freeze({
    intervention_id: contentId({
      applicant_id: applicant.applicant_id,
      check,
      arm,
      targets,
    }),
    family: "MONOTONE" as Family,
    name: `${check}:${arm}`,
    layer: "facts",
    target_field: targetField,
    direction: "set",
    expected_relation: relation,
    params: Object.freeze({ targets: Object.freeze({ ...targets }) }),
  });
}

function arrayEquals(a: readonly string[], b: readonly string[]): boolean {
  if (a.length !== b.length) return false;
  const setA = new Set(a);
  return b.every(x => setA.has(x));
}

function formatPyTuple(arr: readonly string[]): string {
  if (arr.length === 0) return '()';
  if (arr.length === 1) return `('${arr[0]}',)`;
  return `(${arr.map(a => `'${a}'`).join(', ')})`;
}
function buildCase(opts: {
  applicant: Applicant;
  check: string;
  ruleId: string;
  relation: Relation;
  baseTargets: Record<string, any>;
  cfTargets: Record<string, any>;
  expectedBaseBreaches: readonly string[];
  expectedCfBreaches: readonly string[];
}): MonotonicityCase {
  const baseSpec = interventionSpec(opts.applicant, opts.check, "base", opts.baseTargets, opts.relation);
  const cfSpec = interventionSpec(opts.applicant, opts.check, "counterfactual", opts.cfTargets, opts.relation);
  
  const plan = makePairPlan({
    applicant: opts.applicant,
    check: opts.check,
    family: "MONOTONE",
    relation: opts.relation,
    baseArmId: `${opts.check}:base`,
    cfArmId: `${opts.check}:counterfactual`,
    baseInterventions: [baseSpec],
    cfInterventions: [cfSpec],
  });

  const baseApplied = applyInterventions(plan.base.applicant, plan.base.interventions).applicant;
  const cfApplied = applyInterventions(plan.cf.applicant, plan.cf.interventions).applicant;

  const baseEval = evaluate(baseApplied);
  const cfEval = evaluate(cfApplied);

  const baseBreachedIds = baseEval.evaluations.filter((e: any) => e.breached).map((e: any) => e.rule_id);
  const cfBreachedIds = cfEval.evaluations.filter((e: any) => e.breached).map((e: any) => e.rule_id);

  if (!arrayEquals(baseBreachedIds, opts.expectedBaseBreaches)) {
    return inapplicable(
      opts.applicant,
      opts.check,
      opts.ruleId,
      opts.relation,
      `base arm is confounded: expected breached rules ${formatPyTuple(opts.expectedBaseBreaches)}, observed ${formatPyTuple(baseBreachedIds)}`
    );
  }
  if (!arrayEquals(cfBreachedIds, opts.expectedCfBreaches)) {
    return inapplicable(
      opts.applicant,
      opts.check,
      opts.ruleId,
      opts.relation,
      `counterfactual arm is confounded: expected breached rules ${formatPyTuple(opts.expectedCfBreaches)}, observed ${formatPyTuple(cfBreachedIds)}`
    );
  }

  return Object.freeze({
    applicant_id: opts.applicant.applicant_id,
    check: opts.check,
    target_rule_id: opts.ruleId,
    expected_relation: opts.relation,
    pair_id: plan.pair_id,
    cluster_id: plan.cluster_id,
    plan,
    inapplicable_reason: null,
  });
}

function dtiDebtTarget(monthlyIncomeCents: number, ratio: number, above: boolean): number {
  const v = ratio * monthlyIncomeCents;
  return above ? Math.ceil(v) : Math.floor(v);
}

export function buildMonotonicityCases(applicant: Applicant): readonly MonotonicityCase[] {
  const definitions = [
    { check: CHECK_INCOME, ruleId: "min_annual_income", relation: "NONDECREASING" as Relation },
    { check: CHECK_CREDIT_SCORE, ruleId: "min_credit_score", relation: "NONDECREASING" as Relation },
    { check: CHECK_DTI, ruleId: "max_dti", relation: "NONINCREASING" as Relation },
    { check: CHECK_MINOR_DELINQUENCY, ruleId: "max_minor_delinquencies", relation: "NONINCREASING" as Relation },
    { check: CHECK_MAJOR_DELINQUENCY, ruleId: "max_major_delinquencies", relation: "NONINCREASING" as Relation },
  ];

  const sourceEval = evaluate(applicant);
  if (sourceEval.outcome !== "APPROVE") {
    return Object.freeze(definitions.map(d => inapplicable(applicant, d.check, d.ruleId, d.relation, "source profile is not oracle-approved")));
  }

  const incomeRule = policy.rules.find(r => r.rule_id === "min_annual_income")!;
  const scoreRule = policy.rules.find(r => r.rule_id === "min_credit_score")!;
  const dtiRule = policy.rules.find(r => r.rule_id === "max_dti")!;
  const minorRule = policy.rules.find(r => r.rule_id === "max_minor_delinquencies")!;
  const majorRule = policy.rules.find(r => r.rule_id === "max_major_delinquencies")!;

  const incomeLow = Math.max(0, Math.floor(parseFloat(incomeRule.threshold as string) - parseFloat(incomeRule.margin_unit as string)));
  const incomeHigh = Math.floor(parseFloat(incomeRule.threshold as string) + parseFloat(incomeRule.margin_unit as string));

  const scoreLow = Math.max(300, Math.floor(parseFloat(scoreRule.threshold as string) - parseFloat(scoreRule.margin_unit as string)));
  const scoreHigh = Math.floor(parseFloat(scoreRule.threshold as string) + parseFloat(scoreRule.margin_unit as string));

  const dtiLow = parseFloat(dtiRule.threshold as string) - parseFloat(dtiRule.margin_unit as string);
  const dtiHigh = parseFloat(dtiRule.threshold as string) + parseFloat(dtiRule.margin_unit as string);
  
  const monthlyIncomeCents = applicant.facts.annual_income_cents / 12;
  const debtLow = Math.floor(dtiLow * monthlyIncomeCents);
  const debtHigh = Math.ceil(dtiHigh * monthlyIncomeCents);

  const minorLow = Math.floor(parseFloat(minorRule.threshold as string) - parseFloat(minorRule.margin_unit as string));
  const minorHigh = Math.floor(parseFloat(minorRule.threshold as string) + parseFloat(minorRule.margin_unit as string));

  const majorLow = Math.max(0, Math.floor(parseFloat(majorRule.threshold as string)));
  const majorHigh = Math.floor(parseFloat(majorRule.threshold as string) + parseFloat(majorRule.margin_unit as string));

  return Object.freeze([
    buildCase({
      applicant, check: CHECK_INCOME, ruleId: "min_annual_income", relation: "NONDECREASING",
      baseTargets: { annual_income_cents: incomeLow }, cfTargets: { annual_income_cents: incomeHigh },
      expectedBaseBreaches: ["min_annual_income"], expectedCfBreaches: [],
    }),
    buildCase({
      applicant, check: CHECK_CREDIT_SCORE, ruleId: "min_credit_score", relation: "NONDECREASING",
      baseTargets: { credit_score: scoreLow }, cfTargets: { credit_score: scoreHigh },
      expectedBaseBreaches: ["min_credit_score"], expectedCfBreaches: [],
    }),
    buildCase({
      applicant, check: CHECK_DTI, ruleId: "max_dti", relation: "NONINCREASING",
      baseTargets: { monthly_debt_cents: debtLow }, cfTargets: { monthly_debt_cents: debtHigh },
      expectedBaseBreaches: [], expectedCfBreaches: ["max_dti"],
    }),
    buildCase({
      applicant, check: CHECK_MINOR_DELINQUENCY, ruleId: "max_minor_delinquencies", relation: "NONINCREASING",
      baseTargets: { delinq_30d_24m: minorLow, delinq_60d_24m: 0 }, cfTargets: { delinq_30d_24m: minorHigh, delinq_60d_24m: 0 },
      expectedBaseBreaches: [], expectedCfBreaches: ["max_minor_delinquencies"],
    }),
    buildCase({
      applicant, check: CHECK_MAJOR_DELINQUENCY, ruleId: "max_major_delinquencies", relation: "NONINCREASING",
      baseTargets: { delinq_90p_24m: majorLow }, cfTargets: { delinq_90p_24m: majorHigh },
      expectedBaseBreaches: [], expectedCfBreaches: ["max_major_delinquencies"],
    }),
  ]);
}

export function buildAnalyticIncomeCase(applicant: Applicant): MonotonicityCase {
  const definitions = [
    { check: CHECK_INCOME_ANALYTIC, ruleId: "min_annual_income", relation: "NONDECREASING" as Relation }
  ];
  const sourceEval = evaluate(applicant);
  if (sourceEval.outcome !== "APPROVE") {
    return inapplicable(
      applicant,
      CHECK_INCOME_ANALYTIC,
      "min_annual_income",
      "NONDECREASING",
      "source profile is not oracle-approved"
    );
  }
  return buildCase({
    applicant,
    check: CHECK_INCOME_ANALYTIC,
    ruleId: "min_annual_income",
    relation: "NONDECREASING",
    baseTargets: { annual_income_cents: 3_000_000 },
    cfTargets: { annual_income_cents: 4_000_000 },
    expectedBaseBreaches: [],
    expectedCfBreaches: [],
  });
}
